src/meta/meta_service.h

```diff
@@ -414,20 +435,36 @@ bool meta_service::check_status(TRpcHolder rpc, rpc_address *forward_address)
         LOG_INFO("reject request with {}", rpc.response().err);
         return false;
     }
-
     return true;
 }
 
-template <typename TRespType>
-bool meta_service::check_status_with_msg(message_ex *req, TRespType &response_struct)
+// when the Ranger ACL is enabled, only the leader meta_server will pull Ranger policy, so if it is
+// not the leader, _access_controller may be a null pointer, or a new leader is elected, and the
+// above policy information may be out of date.
+template <typename TRpcHolder>
+bool meta_service::check_status_and_authz(TRpcHolder rpc,
+                                          rpc_address *forward_address,
+                                          const std::string &app_name)
 {
-    if (!_access_controller->allowed(req)) {
-        LOG_INFO("reject request with ERR_ACL_DENY");
-        response_struct.err = ERR_ACL_DENY;
-        reply(req, response_struct);
+    if (!check_leader_status(rpc, forward_address)) {
+        return false;
+    }
+    if (!_access_controller->allowed(rpc.dsn_request(), app_name)) {
+        rpc.response().err = ERR_ACL_DENY;
+        LOG_INFO("not authorized {} to operate on app({}) for user({})",
+                 rpc.dsn_request()->rpc_code(),
+                 app_name,
+                 rpc.dsn_request()->io_session->get_client_username());
         return false;
     }
+    return true;
+}
 
+template <typename TRespType>
+bool meta_service::check_status_and_authz_with_reply(message_ex *req,
+                                                     TRespType &response_struct,
+                                                     const std::string &app_name)
+{
     int result = check_leader(req, nullptr);
```

Comment:
Could you please to refactor `check_leader` in another patch firstly to make it return a enum class type? return -1, 0 and 1 is easy to make mistake.